File: python-ref/gvibu_ref/commands/rev.py
"""rev: reverse the characters in each input line."""
from __future__ import annotations

import sys
from pathlib import Path


def reverse_lines(text: str) -> str:
    output: list[str] = []
    for segment in text.splitlines(keepends=True):
        if segment.endswith("\r\n"):
            output.append(segment[-3::-1] + "\r\n")
        elif segment.endswith("\n"):
            output.append(segment[-2::-1] + "\n")
        else:
            output.append(segment[::-1])
    return "".join(output)


def run(args: list[str]) -> int:
    paths: list[str] = []
    parsing_options = True
    for arg in args:
        if parsing_options and arg == "--":
            parsing_options = False
        elif parsing_options and arg.startswith("-"):
            print(f"rev: invalid option: {arg}", file=sys.stderr)
            return 1
        else:
            paths.append(arg)

    if not paths:
        try:
            sys.stdout.write(reverse_lines(sys.stdin.read()))
        except OSError as error:
            print(f"rev: error reading standard input: {error}", file=sys.stderr)
            return 1
        return 0

    exit_code = 0
    for path in paths:
        try:
            text = Path(path).read_text()
            sys.stdout.write(reverse_lines(text))
        except OSError as error:
            print(f"rev: {path}: {error}", file=sys.stderr)
            exit_code = 1
    return exit_code

File: rust/src/commands/rev.rs
//! rev: reverse the characters in each input line.
use std::fs;
use std::io::{self, Read, Write};

fn reverse_lines(input: &str) -> String {
    let mut output = String::with_capacity(input.len());
    for segment in input.split_inclusive('\n') {
        let (body, ending) = if let Some(body) = segment.strip_suffix('\n') {
            if let Some(body) = body.strip_suffix('\r') {
                (body, "\r\n")
            } else {
                (body, "\n")
            }
        } else {
            (segment, "")
        };
        output.extend(body.chars().rev());
        output.push_str(ending);
    }
    output
}

fn write_input(stdout: &mut dyn Write, input: &str) -> i32 {
    if stdout.write_all(reverse_lines(input).as_bytes()).is_ok() {
        0
    } else {
        1
    }
}

pub fn run(stdout: &mut dyn Write, args: &[String]) -> i32 {
    let mut paths = Vec::new();
    let mut parsing_options = true;
    for arg in args {
        if parsing_options && arg == "--" {
            parsing_options = false;
        } else if parsing_options && arg.starts_with('-') {
            eprintln!("rev: invalid option: {}", arg);
            return 1;
        } else {
            paths.push(arg.as_str());
        }
    }

    if paths.is_empty() {
        let mut input = String::new();
        if io::stdin().read_to_string(&mut input).is_err() {
            eprintln!("rev: error reading standard input");
            return 1;
        }
        return write_input(stdout, &input);
    }

    let mut exit_code = 0;
    for path in paths {
        match fs::read_to_string(path) {
            Ok(input) => {
                if write_input(stdout, &input) != 0 {
                    return 1;
                }
            }
            Err(error) => {
                eprintln!("rev: {}: {}", path, error);
                exit_code = 1;
            }
        }
    }
    exit_code
}

#[cfg(test)]
mod tests {
    use super::reverse_lines;

    #[test]
    fn reverses_lines_and_preserves_endings() {
        assert_eq!(reverse_lines("abc\r\nxy"), "cba\r\nyx");
    }
}
